spswarehouse/powerschool/powerschool.py

```diff
@@ -0,0 +1,550 @@
+import os
+
+from ..credentials import powerschool_config
+
+from selenium.webdriver import Chrome
+from selenium.webdriver.chrome import webdriver as chrome_webdriver
+from selenium.webdriver.support.ui import WebDriverWait, Select
+from selenium.webdriver.common.by import By
+from selenium.webdriver.support import expected_conditions as EC
+from selenium.webdriver.common.keys import Keys
+from selenium.webdriver.common.action_chains import ActionChains
+from selenium.common.exceptions import NoSuchElementException
+
+ADMIN_LOGIN_PAGE_PATH = 'admin/pw.html'
+ADMIN_HOME_PAGE_PATH = 'admin/home.html'
+ADMIN_URL_SCHEME = 'https://'
+STATE_REPORTS_PAGE_PATH = 'admin/reports/statereports.html?repType=state'
+REPORT_QUEUE_REPORTWORKS_PAGE_PATH = 'admin/reportqueue/prhome.html'
+REPORT_QUEUE_SYSTEM_PAGE_PATH = 'admin/reportqueue/home.html'
+
+class DriverBuilder:
+    """A set of function used to instantiate a Chrome Selenium Webdriver"""
+    def get_driver(self, download_location=None, headless=False, window_size=(1400, 900),
+                   chrome_option_prefs=None):
+        """
+        Convenience function for creating a chrome driver.
+        :param download_location: A path to where files should be downloaded. Can be absolute or relative.
+        :param headless: A boolean for whether the chromedriver should run without GUI.
+        :param window_size: A tuple of l x w for the browser window
+        :param chrome_option_prefs: A dict() of any options for to apply to the driver using
+        the chrome options class. See http://chromedriver.chromium.org/capabilities and
+        https://chromium.googlesource.com/chromium/src/+/master/chrome/common/pref_names.cc
+        :return: A selenium web driver.
+        """
+
+        driver = self._get_chrome_driver(download_location, headless, chrome_option_prefs)
+
+        driver.set_window_size(*window_size)
+
+        return driver
+
+    def _get_chrome_driver(self, download_location, headless, chrome_option_prefs):
+        chrome_options = chrome_webdriver.Options()
+        prefs = {}
+        if download_location:
+            dl_prefs = {'download.default_directory': os.path.abspath(download_location),
+                        'download.prompt_for_download': False,
+                        'download.directory_upgrade': True,
+                        'safebrowsing.enabled': False,
+                        'safebrowsing.disable_download_protection': True}
+
+            prefs.update(dl_prefs)
+
+        if chrome_option_prefs:
+            prefs.update(chrome_option_prefs)
+        chrome_options.add_experimental_option('prefs', prefs)
+        
+        # when run from a Docker container
+        chrome_options.add_argument('--no-sandbox')
+        chrome_options.add_argument('--disable-dev-shm-usage')
+
+        if headless:
+            chrome_options.add_argument("--headless")
+
+        driver = Chrome(options=chrome_options)
+
+        return driver
+
+class PowerSchool:
```

Comment:
I think it might be helpful to reorder the functions in this class so that the ones you expect an end user to use are at the top, and all the helper functions are below. IIRC, Python doesn't require you to define functions in order because it doesn't check function contents until run time.

In a similar vein, I'm curious how you chose what is and isn't an "internal" function. E.g., the `powerschool_report_helper` functions seem like they should all be internal, but you've designated them as external.

Reply:
Reordered to put the publicly accessible functions at the top. 

My thinking with what to designate as internal is "If someone were using just this class to navigate around PowerSchool to do stuff, would they call this function?" Some users might create a child class like PowerSchoolCALPADS to do things, but others may just prefer to instantiate the class in a Jupyter notebook and call functions one-by-one to get it to do what they want. So basically, everything is public except for functions that are exclusively used by the public functions. I also made get_most_recent_file_in_dir() public, because that was useful in the Airflow pipeline utilities.py file.